File: rust/clikit/src/triage.rs
//! [`Triage`]: the fix directive every diagnostic ends in, per
//! `result-record.schema.json`'s `$defs/triage` and its three closed kinds.

use serde::Serialize;

use crate::error::ClikitError;
use crate::validate;

/// What the caller does next about a diagnostic. Exactly one kind, and the
/// kind fixes which fields are present — the Rust enum makes an
/// out-of-shape combination (a `manual` with a `command`, a `reinvoke` with
/// no `command`) unrepresentable rather than merely schema-invalid.
///
/// Internally tagged on `kind` so the wire form matches the schema exactly:
/// `{"kind":"reinvoke","command":[...]}`, not a nested `{"reinvoke":{...}}`.
#[derive(Debug, Clone, PartialEq, Serialize)]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum Triage {
    /// Run this same CLI again, this way. `command[0]` is this tool.
    Reinvoke {
        /// argv: element 0 is this tool, every other element one already
        /// unquoted argument.
        command: Vec<String>,
        /// Wait this long before retrying. Only meaningful for a transient
        /// failure; absent means retry immediately.
        #[serde(skip_serializing_if = "Option::is_none")]
        after_seconds: Option<u32>,
        /// Why this command fixes the diagnostic, when it isn't obvious
        /// from the diagnostic's own message.
        #[serde(skip_serializing_if = "Option::is_none")]
        instruction: Option<String>,
    },
    /// Run a different executable, this way — the sanctioned exit from the
    /// CLI-before-raw-OS-tool routing rule.
    RunTool {
        /// argv: element 0 is another executable, not this tool.
        command: Vec<String>,
        /// Why this command fixes the diagnostic, when it isn't obvious
        /// from the diagnostic's own message.
        #[serde(skip_serializing_if = "Option::is_none")]
        instruction: Option<String>,
    },
    /// No invocation fixes this; a person must act.
    Manual {
        /// What a person must do, as one line.
        instruction: String,
    },
}

impl Triage {
    /// A `reinvoke` of `command` (argv, element 0 is this tool).
    #[must_use]
    pub fn reinvoke<I, S>(command: I) -> Self
    where
        I: IntoIterator<Item = S>,
        S: Into<String>,
    {
        Triage::Reinvoke {
            command: command.into_iter().map(Into::into).collect(),
            after_seconds: None,
            instruction: None,
        }
    }

    /// A `run_tool` handoff to `command` (argv, element 0 is not this tool).
    #[must_use]
    pub fn run_tool<I, S>(command: I) -> Self
    where
        I: IntoIterator<Item = S>,
        S: Into<String>,
    {
        Triage::RunTool {
            command: command.into_iter().map(Into::into).collect(),
            instruction: None,
        }
    }

    /// A `manual` directive: no invocation fixes this.
    #[must_use]
    pub fn manual(instruction: impl Into<String>) -> Self {
        Triage::Manual {
            instruction: instruction.into(),
        }
    }

    /// Sets the retry floor on a [`Triage::Reinvoke`]. A no-op on
    /// [`Triage::RunTool`] or [`Triage::Manual`], which the schema forbids
    /// it on.
    #[must_use]
    pub fn after_seconds(mut self, seconds: u32) -> Self {
        if let Triage::Reinvoke { after_seconds, .. } = &mut self {
            *after_seconds = Some(seconds);
        }
        self
    }

    /// Attaches an `instruction` to a [`Triage::Reinvoke`] or
    /// [`Triage::RunTool`]. A no-op on [`Triage::Manual`], which already
    /// carries its instruction at construction.
    #[must_use]
    pub fn instruction(mut self, text: impl Into<String>) -> Self {
        match &mut self {
            Triage::Reinvoke { instruction, .. } | Triage::RunTool { instruction, .. } => {
                *instruction = Some(text.into());
            }
            Triage::Manual { .. } => {}
        }
        self
    }

    /// The wire `kind` token.
    #[must_use]
    pub const fn kind(&self) -> &'static str {
        match self {
            Triage::Reinvoke { .. } => "reinvoke",
            Triage::RunTool { .. } => "run_tool",
            Triage::Manual { .. } => "manual",
        }
    }

    pub(crate) fn validate(&self) -> Result<(), ClikitError> {
        match self {
            Triage::Reinvoke {
                command,
                after_seconds,
                instruction,
            } => {
                validate_command(command)?;
                if let Some(instruction) = instruction {
                    validate::validate_line("triage.instruction", instruction, 4096)?;
                }
                if let Some(seconds) = after_seconds {
                    if *seconds > 86400 {
                        return Err(ClikitError::InvalidValue {
                            field: "triage.after_seconds",
                            reason: "must be at most 86400",
                            value: seconds.to_string(),
                        });
                    }
                }
                Ok(())
            }
            Triage::RunTool {
                command,
                instruction,
            } => {
                validate_command(command)?;
                if let Some(instruction) = instruction {
                    validate::validate_line("triage.instruction", instruction, 4096)?;
                }
                Ok(())
            }
            Triage::Manual { instruction } => {
                validate::validate_line("triage.instruction", instruction, 4096)
            }
        }
    }
}

fn validate_command(command: &[String]) -> Result<(), ClikitError> {
    if command.is_empty() || command.len() > 128 {
        return Err(ClikitError::InvalidValue {
            field: "triage.command",
            reason: "must carry 1-128 elements",
            value: format!("{} elements", command.len()),
        });
    }
    for token in command {
        validate::validate_line("triage.command element", token, 4096)?;
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn kind_matches_the_schema_wire_token() {
        assert_eq!(Triage::reinvoke(["x"]).kind(), "reinvoke");
        assert_eq!(Triage::run_tool(["x"]).kind(), "run_tool");
        assert_eq!(Triage::manual("do it").kind(), "manual");
    }

    #[test]
    fn after_seconds_is_a_no_op_off_reinvoke() {
        let triage = Triage::manual("do it").after_seconds(30);
        assert_eq!(triage, Triage::manual("do it"));
    }

    #[test]
    fn empty_command_is_rejected() {
        let empty: Vec<String> = vec![];
        assert!(Triage::reinvoke(empty).validate().is_err());
    }
}

File: go/clikit/triage.go
package clikit

import "fmt"

// TriageKind is a triage directive's closed kind. Adding a member is a MAJOR
// change: a caller matches it exhaustively.
type TriageKind string

const (
	// TriageReinvoke says run this same CLI again, this way. Command[0] is
	// this tool.
	TriageReinvoke TriageKind = "reinvoke"
	// TriageRunTool says run that other executable, this way. Command[0] is
	// not this tool - the sanctioned exit from the CLI-before-raw-OS-tool
	// routing rule.
	TriageRunTool TriageKind = "run_tool"
	// TriageManual says no invocation fixes this; a person must act.
	TriageManual TriageKind = "manual"
)

// Triage is the fix directive every error and caveat carries: what the
// caller does next, executable as given - concrete argv, never a
// placeholder.
type Triage struct {
	Kind         TriageKind `json:"kind"`
	Command      []string   `json:"command,omitempty"`
	Instruction  string     `json:"instruction,omitempty"`
	AfterSeconds int        `json:"after_seconds,omitempty"`
}

// Reinvoke builds a `reinvoke` triage: run this CLI again with command.
func Reinvoke(command ...string) Triage {
	return Triage{Kind: TriageReinvoke, Command: command}
}

// ReinvokeAfter builds a `reinvoke` triage carrying a retry floor - only
// meaningful for a transient failure. after is a floor, not a schedule: the
// earliest a retry is worth attempting.
func ReinvokeAfter(afterSeconds int, command ...string) Triage {
	return Triage{Kind: TriageReinvoke, Command: command, AfterSeconds: afterSeconds}
}

// RunTool builds a `run_tool` triage: hand off to a different executable.
func RunTool(command ...string) Triage {
	return Triage{Kind: TriageRunTool, Command: command}
}

// Manual builds a `manual` triage: no invocation fixes this, instruction says
// what a person does.
func Manual(instruction string) Triage {
	return Triage{Kind: TriageManual, Instruction: instruction}
}

// validate checks t against the triage shape and kind-specific rules the
// schema pins.
func (t Triage) validate() error {
	switch t.Kind {
	case TriageReinvoke:
		if len(t.Command) == 0 {
			return fmt.Errorf("clikit: reinvoke triage requires command")
		}
	case TriageRunTool:
		if len(t.Command) == 0 {
			return fmt.Errorf("clikit: run_tool triage requires command")
		}
		if t.AfterSeconds != 0 {
			return fmt.Errorf("clikit: run_tool triage must not carry after_seconds")
		}
	case TriageManual:
		if t.Instruction == "" {
			return fmt.Errorf("clikit: manual triage requires instruction")
		}
		if len(t.Command) != 0 {
			return fmt.Errorf("clikit: manual triage must not carry command")
		}
		if t.AfterSeconds != 0 {
			return fmt.Errorf("clikit: manual triage must not carry after_seconds")
		}
	default:
		return fmt.Errorf("clikit: unknown triage kind %q", string(t.Kind))
	}
	if len(t.Command) > 128 {
		return fmt.Errorf("clikit: triage command has %d elements, max 128", len(t.Command))
	}
	for _, tok := range t.Command {
		if !isArgvToken(tok) {
			return fmt.Errorf("clikit: invalid triage command token %q", tok)
		}
	}
	if t.Instruction != "" && !isLine(t.Instruction) {
		return fmt.Errorf("clikit: invalid triage instruction %q", t.Instruction)
	}
	if t.AfterSeconds < 0 || t.AfterSeconds > 86400 {
		return fmt.Errorf("clikit: after_seconds %d out of range [0,86400]", t.AfterSeconds)
	}
	return nil
}
